sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveTableScanExec.scala

```diff
@@ -0,0 +1,501 @@
+                                             maxReaderBatchSizeBytes: Long,
+                                             metrics: Map[String, GpuMetric],
+                                             @transient options: Map[String, String])
+  extends ShimFilePartitionReaderFactory(options) {
+
+  override def buildReader(partitionedFile: PartitionedFile): PartitionReader[InternalRow] = {
+    throw new IllegalStateException("Row-based text parsing is not supported on GPU.")
+  }
+
+  private val csvOptions = new CSVOptions(options,
+                                          sqlConf.csvColumnPruning,
+                                          sqlConf.sessionLocalTimeZone,
+                                          sqlConf.columnNameOfCorruptRecord)
+
+  override def buildColumnarReader(partFile: PartitionedFile): PartitionReader[ColumnarBatch] = {
+    val conf = broadcastConf.value.value
+    val reader = new PartitionReaderWithBytesRead(
+                   new GpuHiveDelimitedTextPartitionReader(
+                     conf, csvOptions, partFile, inputFileSchema,
+                     requestedOutputDataSchema, maxReaderBatchSizeRows,
+                     maxReaderBatchSizeBytes, metrics))
+
+    val partValueScalars = ColumnarPartitionReaderWithPartitionValues.createPartitionValues(
+      partFile.partitionValues.toSeq(partitionSchema),
+      partitionSchema
+    )
+    new AlphabeticallyReorderingColumnPartitionReader(reader,
+                                                      partValueScalars,
+                                                      GpuColumnVector.extractTypes(partitionSchema),
+                                                      partitionSchema,
+                                                      requestedAttributes)
+  }
+}
+
+// Reader that converts from chunked data buffers into cudf.Table.
+class GpuHiveDelimitedTextPartitionReader(conf: Configuration,
+                                          csvOptions: CSVOptions,
+                                          partFile: PartitionedFile,
+                                          inputFileSchema: StructType,
+                                          requestedOutputDataSchema: StructType,
+                                          maxRowsPerChunk: Integer,
+                                          maxBytesPerChunk: Long,
+                                          execMetrics: Map[String, GpuMetric]) extends
+  CSVPartitionReader(conf, partFile, inputFileSchema, requestedOutputDataSchema,
+                     csvOptions, maxRowsPerChunk, maxBytesPerChunk, execMetrics) {
+
+  override def buildCsvOptions(parsedOptions: CSVOptions,
+                               schema: StructType,
+                               hasHeader: Boolean): ai.rapids.cudf.CSVOptions.Builder = {
+    super.buildCsvOptions(parsedOptions, schema, hasHeader)
+      .withDelim('\u0001') // Record field delimiter '^A'.
```

Comment:
What about the line separator? Is the default for hive really just "\n"? If so then this differs from spark that has the default be "\n" or "\r" or "\r\n". The issue is with the HadoopTextLinesReader. In CUDF it does the right thing, but I want to be sure that if a "\r" shows up in the wrong place in a hive file that we parse it correctly.

Reply:
I haven't tested this with `\r` or `\r\n`. It doesn't look like Spark supports reading from Hive text tables with a custom line separator:
```
scala> sql(" create table ctrl_r(foo string) row format delimited lines terminated by '\r' stored as textfile location '/tmp/ctrl_r' ").show(false)
org.apache.spark.sql.catalyst.parser.ParseException:
(line 1, pos 33) BY only supports newline '\n' right now:

== SQL ==
' stored as textfile location '/tmp/ctrl_r' delimited lines terminated by '
---------------------------------^^^

  at org.apache.spark.sql.catalyst.parser.ParserUtils$.validate(ParserUtils.scala:120)
  at org.apache.spark.sql.catalyst.parser.AstBuilder.$anonfun$visitRowFormatDelimited$3(AstBuilder.scala:3241)  at scala.collection.immutable.List.map(List.scala:293)  at org.apache.spark.sql.catalyst.parser.AstBuilder.$anonfun$visitRowFormatDelimited$1(AstBuilder.scala:3236)
  at org.apache.spark.sql.catalyst.parser.ParserUtils$.withOrigin(ParserUtils.scala:133)
  at org.apache.spark.sql.catalyst.parser.AstBuilder.visitRowFormatDelimited(AstBuilder.scala:3223)
...
```

Reply:
What I am more concerned about is that that we are reusing the CSV code to read the hive text table. The CSV code by default will treat a file that looks like "A\rB\rC" the same as "A\nB\nC" the same as a file that looks like "A\rB\nC\r\n" I don't think we want that for hive files. What I am suggesting is the CSVOptions that are passed into this class need to explicitly set the line separator so that it will only look for "\n" instead of what CSV does by default.

Reply:
That's a good catch. I'll test this to confirm, and modify the `CSVOptions`.

Reply:
I have verified that Hive's `LazySimpleSerDe` treats both `\r` and `\n` identically: they both function as the default record delimiter.
I have also verified that in its current state, the change introduced here also functions similarly. That's the good news.

However, there is a corner case. Consider the following input file (with a single column), where:
1. `\r` is represented by `^M`
2. `\n` is represented by a new line

```
123^M
456^M
```
With the plugin *disabled*, `LazySimpleSerDe` treats this as *4* rows:
```
+---+
|str|
+---+
|123|
|   |
|456|
|   |
+---+
```
With the plugin *enabled*, empty lines are *skipped*, yielding only *2* rows:
```
+---+
|str|
+---+
|123|
|456|
+---+
```
This is because `Table.readCSV()` drops blank lines by default. We don't currently have JNI access to [`cudf::io::csv_reader_options::enable_skip_blank_lines()`](https://github.com/rapidsai/cudf/blob/branch-22.12/cpp/include/cudf/io/csv.hpp#L665).

I'll file an issue for this.

Reply:
https://github.com/NVIDIA/spark-rapids/issues/7068